XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,628 @@
+assigned task has either completed or encountered an error. Requests are used
+by handles to perform work. Such as accepting a new connection or writing data
+to disk.
+
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// Return the id of the current execution context. Useful for tracking state
+// and retrieving the handle of the current parent without needing to use an
+// AsyncHook.
+const id = async_wrap.currentId();
+
+// Return new AsyncEvent instance. Used to trigger the AsyncHook callbacks.
+const asyncEvent = async_wrap.createEvent(handle[, parent_id]);
+
+// Returns the id assigned to the AsyncEvent instance.
+const id = asyncEvent.getId();
+
+// Trigger init() callbacks.
+asyncEvent.emitInit();
+
+// Trigger before() callbacks.
+asyncEvent.emitBefore();
+
+// Trigger after() callbacks.
+asyncEvent.emitAfter();
+
+// Trigger destroy() callbacks.
+asyncEvent.emitDestroy();
+
+// Create a new instance and add each callback to the corresponding hook.
+const asyncHook = new async_wrap.createHook({ init, before, after, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.syncScope();
+
+// Unlike disable(), prevent any hook callback from firing again in the future.
+// Future hooks can be added by running syncScope()/enable() again.
+// TODO: Investigation is needed to know whether this is possible.
+asyncHook.remove();
+
+// init() is called during object construction. The handle may not have
+// completed construction when this callback runs. So all fields of the
+// handle referenced by "id" may not have been populated.
+function init(id, type, parentId, handle) { }
+
+// before() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWrap), and should be called exactly 1
+// time for requests (e.g. FSReqWrap).
+function before(id) { }
+
+// after() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw.
+function after(id) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPParser where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after after() has
+// completed.
+function destroy(id) { }
```

Comment:
@RobinQu You don't need Async_Wrap to implement Zones in your own code, maybe from other APIs.

If you do `process.binding('async_wrap')` it will not match this document.

The API described in this document is not yet in a Node release. Please wait.